packages/@lwc/template-compiler/src/codegen/index.ts

```diff
@@ -205,18 +205,30 @@ function transform(codeGen: CodeGen): t.Expression {
      * Transforms an IfBlock or ElseifBlock along with both its direct descendants and its 'else' descendants.
      *
      * @param conditionalParentBlock The IfBlock or ElseifBlock to transform into a conditional expression
+     * @param key The key to use for this chain of IfBlock/ElseifBlock branches, if applicable
      * @returns A conditional expression representing the full conditional tree with conditionalParentBlock as the root node
      */
     function transformConditionalParentBlock(
-        conditionalParentBlock: IfBlock | ElseifBlock
+        conditionalParentBlock: IfBlock | ElseifBlock,
+        key?: number
     ): t.Expression {
-        const childrenExpression = transformChildren(conditionalParentBlock);
-        let elseExpression: t.Expression = t.arrayExpression([]);
+        const ifBlockKey = key || codeGen.generateKey();
 
+        const childrenExpression = codeGen.genFragment(
+            t.literal(ifBlockKey),
+            transformChildren(conditionalParentBlock)
```

Comment:
`genFragment` has a third parameter that lets us specify whether this fragment is `stable` or not. A subsequent perf optimization we can do is to detect whether the children here are dynamic or static and set the `stable` flag appropriately.

Leaving that for a separate PR after https://github.com/salesforce/lwc/pull/3030 gets merged. Getting that merged sooner will unblock some of the integration work.